File: src/setup/QuestionBuilder.php
<?php

namespace pocketcloud\cloud\setup;

use Closure;

final class QuestionBuilder {

    private Closure $parser;
    private bool $canSkipped = false;
    private array $possibleAnswers = [];
    private ?string $defaultValueMessage = null;
    private mixed $actualDefaultValue = null;
    private ?string $recommendation = null;
    private ?Closure $resultHandler = null;

    public static function builder(string $id, string $question): QuestionBuilder {
        return new self($id, $question);
    }

    public function __construct(
        private readonly string $id,
        private readonly string $question
    ) {
        $this->parser = self::defaultParser();
    }

    public function canSkipped(bool $value): self {
        $this->canSkipped = $value;
        return $this;
    }

    public function possibleAnswers(string|int|float|bool ...$answers): self {
        $this->possibleAnswers = $answers;
        return $this;
    }

    public function default(string $displayDefault, mixed $value): self {
        $this->defaultValueMessage = $displayDefault;
        $this->actualDefaultValue = $value;
        return $this;
    }

    public function recommendation(string $recommendation): self {
        $this->recommendation = $recommendation;
        return $this;
    }

    /**
     * @param Closure(string $input, ?string &$error): mixed $parser
     * @return $this
     * @required
     */
    public function parser(Closure $parser): self {
        $this->parser = $parser;
        return $this;
    }

    public function resultHandler(Closure $handler): self {
        $this->resultHandler = $handler;
        return $this;
    }

    public function build(): Question {
        return new Question(
            $this->id,
            $this->question,
            $this->canSkipped,
            $this->possibleAnswers,
            $this->defaultValueMessage,
            $this->actualDefaultValue,
            $this->recommendation,
            $this->parser,
            $this->resultHandler
        );
    }

    public static function defaultParser(): Closure {
        return fn(string $input): string => $input;
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/setup/QuestionBuilder.java
package de.pocketcloud.cloud.setup;

import java.util.Arrays;
import java.util.List;
import java.util.function.Consumer;

public final class QuestionBuilder<T> {

    private final String id;
    private final String question;
    private boolean canSkipped = false;
    private List<String> possibleAnswers = List.of();
    private String defaultValueMessage;
    private T defaultValue;
    private String recommendation;
    private final QuestionParser<T> parser;
    private Consumer<T> resultHandler;

    private QuestionBuilder(String id, String question, QuestionParser<T> parser) {
        this.id = id;
        this.question = question;
        this.parser = parser;
    }

    public static QuestionBuilder<String> builder(String id, String question) {
        return new QuestionBuilder<>(id, question, defaultParser());
    }

    public QuestionBuilder<T> canSkipped(boolean value) {
        this.canSkipped = value;
        return this;
    }

    public QuestionBuilder<T> possibleAnswers(String... answers) {
        this.possibleAnswers = Arrays.asList(answers);
        return this;
    }

    public QuestionBuilder<T> possibleAnswers(List<String> answers) {
        this.possibleAnswers = answers;
        return this;
    }

    public QuestionBuilder<T> defaultValue(String displayDefault, T value) {
        this.defaultValueMessage = displayDefault;
        this.defaultValue = value;
        return this;
    }

    public QuestionBuilder<T> recommendation(String recommendation) {
        this.recommendation = recommendation;
        return this;
    }

    public <R> QuestionBuilder<R> parser(QuestionParser<R> parser) {
        QuestionBuilder<R> next = new QuestionBuilder<>(this.id, this.question, parser);
        next.canSkipped = this.canSkipped;
        next.possibleAnswers = this.possibleAnswers;
        next.recommendation = this.recommendation;
        return next;
    }

    public QuestionBuilder<T> resultHandler(Consumer<T> handler) {
        this.resultHandler = handler;
        return this;
    }

    public Question<T> build() {
        return new Question<>(
                id,
                question,
                canSkipped,
                possibleAnswers,
                defaultValueMessage,
                defaultValue,
                recommendation,
                parser,
                resultHandler
        );
    }

    public static QuestionParser<String> defaultParser() {
        return (input, error) -> input;
    }
}